server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocatorTests.java

```diff
@@ -333,6 +336,73 @@ public ShardAllocationDecision decideShardAllocation(ShardRouting shard, Routing
         assertThat(listenersCalled.get(), equalTo(1));
     }
 
+    public void testConcurrency() throws Exception {
+
+        var createdIndices = new CopyOnWriteArraySet<String>();
+        var allocatedIndices = new CopyOnWriteArraySet<String>();
+        var reroutedIndices = new CopyOnWriteArraySet<String>();
+
+        var threadPool = new TestThreadPool(getTestName());
+        RerouteService rerouteService = (r, p, l) -> threadPool.executor(ThreadPool.Names.CLUSTER_COORDINATION).submit(() -> {
+            reroutedIndices.addAll(createdIndices);
+            l.onResponse(null);
+        });
```

Comment:
Think it's slightly stronger to capture the created indices earlier here? Also `ThreadPool.Names.CLUSTER_COORDINATION` seems like the wrong pool (and it only has one thread), this should work with `GENERIC` too:

```
        RerouteService rerouteService = (r, p, l) -> {
            final var localCreatedIndices = Set.copyOf(createdIndices);
            threadPool.executor(ThreadPool.Names.GENERIC).submit(() -> {
                reroutedIndices.addAll(localCreatedIndices);
                l.onResponse(null);
            });
        };
```